SU2_CFD/src/solvers/CDiscAdjSolver.cpp

```diff
@@ -123,6 +123,8 @@ void CDiscAdjSolver::SetRecording(CGeometry* geometry, CConfig *config){
   }
   END_SU2_OMP_FOR
 
+  direct_solver->GetNodes()->SetSolutionExtra(nodes->GetSolutionExtra_Direct());
```

Comment:
Instead of keeping and passing around vectors of pointers, make SetSolutionExtra a method of the solver, then it can set the values of its extra variables directly.

Reply:
I am unsure how it helps if the the method moves to the solver because the fundamental problem is:

`SolutionExtra_OriginAdresses, AD_InputIndex_Extra, AD_OutputIndex_Extra` are owned by the direct_solver

`SolutionExtra, ExternalExtra, SolutionExtra_BGS_k, SolutionExtra_Direct` are owned by the DiscAdj solver.

So I need to pass `SolutionExtra_Direct` from the DiscAdj solver to the direct_solver such that the direct_solver can reset the primal values at the `SolutionExtra_OriginAdresses`. And I need calls to a Variable method anyway as the values are at least protected, meaning: a solver method would add code without making something else more concise?

Reply:
By the way -> This distinction of container-ownership between direct and adjoint solver is already present for the current DA solver. 

With the only slight difference being `SolutionExtra_Direct` where the "normal" counterpart is part of `CDiscAdjVariable` and not `CVariable`... which saves me some virtuals

Reply:
> I told you in the meeting to not put this in CVariable but here we are...

I suggest to apply [Hanlon's razor](https://en.wikipedia.org/wiki/Hanlon%27s_razor) when it comes to my work... and this post

> Put these in the direct solver:
> SolutionExtra_OriginAdresses, AD_InputIndex_Extra, AD_OutputIndex_Extra

I guess by direct solver you mean FVMFlowSolverBase, which would make the implementation not generic for HeatSolver and ScalarSolver(s) which are a direct children of CSolver. So in case someone needs to do this there as well, code has to be copied, which I am unsure makes anything better.

> Which means that SolutionExtra_OriginAdresses is not needed anymore since the original variables belong to the solver already.

That means that if I have a variable I want to use as an additional adjoint one needs to manually handle it for storing and restoring the saved value plus at RegisterIn/Output. Storing the address happens once (preferably while in the constructor) and everything else is done automatically. 

> Then what is left in CVariable belongs to the DA solver, which makes the separation clearer.

Fully agree on the clearer thing. Currently it is not obvious from looking at CVariable where each container belongs to. The DAVariable or solver is for all solvers so moving things there does not have the same problems with generality as the direct_solver has.


> You don't need to make this look like the rest of the code.
> For example, the DA solver never needed to know about "SolutionDirect", it should have simply asked the primal to reset its solution, whatever that solution is.

I agree, makes more sense to have SolutionDirect stored in the primal solver. But keeping things consistent is always nice, big fan.

Reply:
You don't need to have those actual fields in csolver, what you need is an interface in csolver that's allows you to implement the actions you need.
Registering extra variables, restoring values, getting the adjoints, setting the adjoints etc.
You can even dispense with the index bookkeeping and reset if you don't modify the starting value, and instead write the updated value to another variable (in adjoint mode) that way you have one input and one output and no need to store indices.

Reply:
RegisterExtraVariables (gives you a count for you to resize the extra stuff in the DA solver cvariable)
SetAdjointExtraVariable(vector)
GetDerivativeExtraVariable -> vector